torch/csrc/jit/pickler.cpp

```diff
@@ -0,0 +1,391 @@
+}
+
+void Unpickler::run() {
+  // Expect a PROTO opcode and protocol number at the start of blob
+  AT_ASSERT(readOpCode() == OpCode::PROTO);
+  uint8_t protocol = read<uint8_t>();
+  AT_CHECK(
+      protocol == 2,
+      "Only Pickle protocol 2 is supported, found protocol = ",
+      protocol);
+
+  while (bytes_ < end_ptr_) {
+    OpCode opcode = readInstruction();
+    if (opcode == OpCode::STOP) {
+      return;
+    }
+  }
+
+  AT_ERROR("Overran buffer while unpickling data, didn't find STOP opcode");
+}
+
+OpCode Unpickler::readInstruction() {
+  auto op = readOpCode();
+  switch (op) {
+    case OpCode::EMPTY_LIST: {
+      // TODO: Use fake classes to mark list specializations
+      stack_.emplace_back(std::vector<IValue>());
+      break;
+    }
+    case OpCode::EMPTY_TUPLE: {
+      stack_.emplace_back(c10::ivalue::Tuple::create({}));
+      break;
+    }
+    case OpCode::BINPUT: {
+      size_t memo_id = read<uint8_t>();
+      if (memo_.size() <= memo_id) {
+        memo_.reserve(1 + 2 * memo_id);
+      }
+      memo_.push_back(stack_.back());
+      break;
+    }
+    case OpCode::MARK: {
+      // Mark location of the container ivalue in the stack
+      marks_.push_back(stack_.size());
+      break;
+    }
+    case OpCode::BININT: {
+      int32_t value = read<uint32_t>();
+      stack_.emplace_back(int64_t(value));
+      break;
+    }
+    case OpCode::BINUNICODE: {
+      int32_t length = read<uint32_t>();
+      const char* characters = reinterpret_cast<const char*>(bytes_);
+      bytes_ += length;
+      stack_.emplace_back(std::string(characters, /*n=*/length));
+      break;
+    }
+    case OpCode::BINFLOAT:
+      stack_.emplace_back(readFloat());
+      break;
+    case OpCode::TUPLE: {
+      size_t start = marks_.back();
+      marks_.pop_back();
+      IValue tup = c10::ivalue::Tuple::create(
+          std::vector<IValue>(stack_.begin() + start, stack_.end()));
+      stack_.resize(start);
+      stack_.push_back(tup);
+    } break;
+    case OpCode::EMPTY_DICT:
+      stack_.emplace_back(c10::ivalue::UnorderedMap());
+      break;
+    case OpCode::APPENDS: {
+      size_t start = marks_.back();
+      marks_.pop_back();
+      auto list = stack_[start - 1].toGenericList();
+      list->elements().insert(
+          list->elements().end(), stack_.begin() + start, stack_.end());
+      stack_.resize(start);
+    } break;
+    case OpCode::SETITEMS: {
+      size_t start = marks_.back();
+      marks_.pop_back();
+      auto dict = stack_[start - 1].toGenericDict();
+      for (size_t i = start; i < stack_.size(); i += 2) {
+        dict->elements()[stack_[i]] = stack_[i + 1];
+      }
+      stack_.resize(start);
+    } break;
+    case OpCode::BINGET: {
+      stack_.push_back(memo_.at(read<uint8_t>()));
+    } break;
+    case OpCode::STOP:
+      return op;
+    case OpCode::GLOBAL: {
+      AT_ASSERT(readString() == "__main__");
+      // Push class name to stack
+      stack_.emplace_back(readString());
```

Comment:
It is better to decode the class name here into an enum and push an integer. This op will happen once per pickle blob, but the `BUILD` instruction will happen once per Tensor. This would replace a bunch of string hashing with a switch statement on 3 entries.